### ISSUE TITLE
Napalm / Capirca Issue 

### ISSUE THREAD

**Issue by prashanthtuttu** [saltstack/salt#47117]
### Description of Issue/Question
Napalm / Capirca Issue

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
As suggested by **mirceaulinic** Copied the capirca_acl.py module to _module directory, but still facing the below error
https://github.com/saltstack/salt/issues/47011

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)
testdevice-ios:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1493, in _thread_return
        return_data = executor.execute()
      File "/usr/lib/python2.7/dist-packages/salt/executors/direct_call.py", line 28, in execute
        return self.func(*self.args, **self.kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/utils/napalm.py", line 430, in func_wrapper
        return func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/napalm_acl.py", line 643, in load_filter_config
        filter_config = __salt__['capirca.get_filter_config'](platform,
      File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1114, in __getitem__
        func = super(LazyLoader, self).__getitem__(item)
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 101, in __getitem__
        raise KeyError(key)
    KeyError: 'capirca.get_filter_config'

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
Salt Version:
           Salt: 2017.7.3

Dependency Versions:
           cffi: 1.11.4
       cherrypy: 3.5.0
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Dec  4 2017, 14:50:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-87-generic
         system: Linux
        version: Ubuntu 16.04 xenial

**Comment by Ch3LL** [comment]
can you see latest comment from @mirceaulinic from here: https://github.com/saltstack/salt/issues/47011#issuecomment-381985137 and paste asnwer here

**Comment by prashanthtuttu** [comment]
Attaching the result
https://gist.github.com/prashanthtuttu/80586187371273188b44946c65b95609

**Comment by mirceaulinic** [comment]
Yes, that looks good. As longs as you started the process for the Proxy Minion after you installed Capirca, and follow the documentation notes from https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.capirca_acl.html, everything should be alright.

**Comment by prashanthtuttu** [comment]
@mirceaulinic - i Followed the same steps mentioned on th link but still issue exists. i run the below cli command
salt 'device2' capirca.get_filter_config junos my-filter pillar_key=netacl test=true  and get same error on both version.

device2:
    'capirca.get_filter_config' is not available.

**Comment by mirceaulinic** [comment]
@prashanthtuttu is the Proxy process running on the same machine as your Master?

**Comment by prashanthtuttu** [comment]
@mirceaulinic - I have 2 setup , 1. master and proxy in same server , 2. Both in diff server. 
Both the cases i face same issues

**Comment by mirceaulinic** [comment]
The recipe is simple: install Capirca as documented wherever you have the Salt Proxy Minion running, and make sure to restart the Proxy process after restart. That's all.

**Comment by prashanthtuttu** [comment]
@mirceaulinic - Followed the steps mentioned in the link https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.capirca_acl.html and restarted the proxy minion process , but still getting the error.

testdevice-junos:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1600, in _thread_return
        return_data = minion_instance.executors[fname](opts, data, func, args, kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/executors/direct_call.py", line 12, in execute
        return func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/utils/napalm.py", line 430, in func_wrapper
        return func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/napalm_acl.py", line 433, in load_term_config
        term_config = __salt__['capirca.get_term_config'](platform,
      File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1142, in __getitem__
        func = super(LazyLoader, self).__getitem__(item)
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 104, in __getitem__
        raise KeyError(key)
    KeyError: u'capirca.get_term_config'

**Comment by mirceaulinic** [comment]
Can you execute the following commands from the Master:

```bash
$ sudo salt testdevice-junos test.versions_report
$ sudo salt testdevice-junos pip.freeze
```

And share their output?

**Comment by prashanthtuttu** [comment]
@mirceaulinic - pls find the outputs in 
https://gist.github.com/prashanthtuttu/50bb9435578f9cc597782a3bbd2362dc

**Comment by mirceaulinic** [comment]
@prashanthtuttu at a glance that seem to be alright, though I'm not sure what could possible cause that. Try uninstalling `napalm-base` (and any other `napalm-*` library), though that shouldn't make any difference as you're running `napalm>2.0`.

Maybe it would help if you'd play with a proper environment, and check the differences yourself:

1. Clone https://github.com/mirceaulinic/salt-docker/tree/salt-47117 (and use the `salt-47117` branch).
2. Adjust the data in `pillar/napalm_pillar.sls` (https://github.com/mirceaulinic/salt-docker/blob/salt-47117/pillar/napalm_pillar.sls) to start your Proxy properly.
3. Execute `make PROXYID=napalm start`:

```bash
$ make PROXYID=napalm start                                                                                                                                                                              
docker-compose up -d
Creating salt-master ...
Creating salt-master ... done
Creating salt-proxy-napalm ...
Creating salt-proxy-napalm ... done
```

4. Execute the commands: https://gist.github.com/mirceaulinic/9df5da1b0163e7d64b8102d90183df0b

**Comment by mirceaulinic** [comment]
@prashanthtuttu it's very strange that in https://github.com/saltstack/salt/issues/46889#issuecomment-379162773 it seems that you managed to install Capirca and all the dependencies correctly. Would you be able to clarify what are you doing different now?

**Comment by prashanthtuttu** [comment]
@mirceaulinic - Thanks , I followed the steps and was able to see things working better than the past.
but still facing a small issue like - when i run >
 salt 'napalm' capirca.get_policy_config juniper pillar_key=netacl - able to get the result but
when running 
salt napalm netacl.load_term_config my-filter term_name=my-term pillar_key=netacl test=True debug=True  - Error is
napalm:
    'netacl.load_term_config' is not available.

Am i missing something here?

**Comment by mirceaulinic** [comment]
```
salt napalm netacl.load_term_config my-filter term_name=my-term pillar_key=netacl test=True debug=True - Error is
napalm:
'netacl.load_term_config' is not available.
```

Yes, this one is a bug indeed. Thanks!

**Comment by mirceaulinic** [comment]
Submitted https://github.com/saltstack/salt/pull/47241 for the fix.
@prashanthtuttu I have also ported the modules and patched into my `salt-docker` repo, please go ahead and test if #47241 does indeed fix the issue.

**Comment by prashanthtuttu** [comment]
Thank you @mirceaulinic , its works fine now , let me compare this setup with my setup and figure out what did i miss and update you.

**Comment by rallytime** [comment]
Closed via #47241
